formula/src/main/java/com/instacart/formula/internal/SynchronizedUpdateQueue.kt

```diff
@@ -0,0 +1,116 @@
+package com.instacart.formula.internal
+
+import java.util.concurrent.ConcurrentLinkedQueue
+import java.util.concurrent.atomic.AtomicReference
+
+/**
+ * We can only process one formula update at a time. To enable thread-safety we use a
+ * non-blocking event queue with a serial confinement strategy for queue processing. All external
+ * formula events use [postUpdate] function which adds the update to [updateQueue], and then,
+ * tries to start processing the queue via atomic [threadRunning] variable. If another thread
+ * was first to take over [threadRunning], we let that thread continue and we exit out. Given
+ * all formula state access is gated via [threadRunning] atomic reference, we are able to ensure
+ * that there is happens-before relationship between each thread and memory changes are visible
+ * between them.
+ */
+class SynchronizedUpdateQueue {
+    /**
+     * Defines a thread currently executing formula update. Null value indicates idle queue.
+     *
+     * To ensure that memory changes within formula internals are synchronized between threads,
+     * we piggyback on the internal synchronization of this variable. Modification to this
+     * variable wraps around every formula update:
+     * - threadRunning = MyThread
+     * - formulaUpdate()
+     * - threadRunning = null
+     *
+     * This creates happens-before relationship between multiple threads and makes sure that
+     * all modifications within formulaUpdate() block are visible to the next thread.
+     */
+    private val threadRunning = AtomicReference<Thread>()
+
+    /**
+     * A non-blocking thread-safe FIFO queue that tracks pending updates.
+     */
+    private val updateQueue = ConcurrentLinkedQueue<() -> Unit>()
+
+    /**
+     * To ensure that we execute one update at a time, all external formula events use this
+     * function to post updates. We add the update to a queue and then try to start processing.
+     * Failure to start processing indicates that another thread was first and we allow that
+     * thread to continue.
+     */
+    fun postUpdate(update: () -> Unit) {
+        val currentThread = Thread.currentThread()
+        val owner = threadRunning.get()
+        if (owner == currentThread) {
+            // This indicates a nested update where an update triggers another update. Given we
+            // are already thread gated, we can execute this update immediately without a need
+            // for any extra synchronization.
+            update()
+            return
+        }
+
+        val success = if (updateQueue.peek() == null) {
+            // No pending update, let's try to run our update immediately
+            takeOver(currentThread) { update() }
```

Comment:
nit: `takeOver(currentThread) { update() }` -> `takeOver(currentThread, update)` to avoid anonymous lambda?

Reply:
could be impossible because `takeOver` is inline

Reply:
Removed inline function and updated to `takeOver(currentThread, update)`